src/core/inductive.ml

```diff
@@ -0,0 +1,244 @@
+         then the function [domrec a x rec_hyp next] is applied to conclude
+         the building. ([rec_hyp] is the current recursive hypothesis and
+         [next] is the result of the recursive call.
+      3) If the current type is a product [Π(x:a), b] not of the previous
+         form, then the function [dom a x next] is called.
+      4) Any other case is an error.*)
+let fold_cons_typ (pos : popt) (codom : term list -> 'b list -> 'a -> 'c)
+      (inj_var : 'b list -> tvar -> 'b)
+      (build_rec_hyp : term list -> 'b -> 'a)
+      (domrec : term -> 'b -> 'a -> 'c -> 'c) (dom : term -> 'b -> 'c -> 'c)
+      (ind_sym : sym) (cons_sym : sym) (f_rec : 'a -> 'b -> 'a -> 'a)
+      (f_not_rec : 'a -> 'b -> 'a) (init : 'a) (s : string) : 'c =
+  let rec aux : 'b list -> 'a -> term -> 'c = fun xs acc a ->
+    match Basics.get_args a with
+    | (Symb(s), ts) ->
+       if s == ind_sym then codom ts xs acc
+       else fatal pos "%a is not a constructor of %a"
+              pp_symbol cons_sym pp_symbol ind_sym
+    | (Prod(a,b), _) ->
+        let (x,b) = Basics.unbind_name b s in
+        let x = inj_var xs x in
+       begin
+         match Basics.get_args a with
+         | (Symb(s), ts) ->
+             if s == ind_sym then
+               let rec_hyp = build_rec_hyp ts x in
+               let next = aux (x::xs) (f_rec acc x rec_hyp) b in
+               domrec a x rec_hyp next
+             else
+               let next = aux (x::xs) (f_not_rec acc x) b in
+               dom a x next
+         | _ -> fatal pos "The type of %a is not supported" pp_symbol cons_sym
+       end
+    | _ -> fatal pos "The type of %a is not supported" pp_symbol cons_sym
+  in aux [] init !(cons_sym.sym_type)
+
+(** [gen_rec_type ss pos ind_sym cons_list] generates the induction principle
+    of the inductive type [ind_sym] (and its position [pos]) with the cons-
+    tructors [cons_list]. Each recursive argument is followed by its induction
+    hypothesis. For instance, with [inductive T:TYPE := c: T->T->T], we have
+    [ind_T: Πp:T->Prop, (Πx0:T, π(p x0)-> Πx1:T, π(p x1)-> π(p (c x0 x1)) ->
+    Πx:T, π(p x)]. Indeed, if the user doesn't give a name for an argument
+    (in case of no dependent product in fact), a fresh name will create (xi
+    with a fresh i). In general, thanks to this function, the command
+    inductive T : Π(i1:A1),...,Π(im:Am), TYPE := c1:T1 | ... | cn:Tn generates
+    ind_T: Π(p:Π(i1:A1),...,Π(im:Am), T i1 ... im -> Prop), U1 -> ... -> Un ->
+    (Π(i1:A1),...,Π(im:Am), Π(t:T i1 ... im), π(p i1 ... im t))
+    where Ui is the clause associated to the constructor ci. *)
+let gen_rec_type : Sig_state.t -> popt -> sym -> sym list -> term =
+  fun ss pos ind_sym cons_list ->
+
+  (* STEP 0: Define some tools which will be useful *)
+  let c = get_config ss pos in
+  let l = List.map (fun e -> e.sym_name) (ind_sym::cons_list) in
+  let set = Extra.StrSet.of_list l in
```

Comment:
This can be improved by recording constructor names starting by "x" and "p" only:
```
let set =
  let f set sym =
    let s = sym.sym_name in
    if s <> "" && (s.[0] = 'x' || s.[0] = 'p') then StrSet.add s set else set
  in 
  List.fold_left f StrSet.empty (ind_sym::cons_list)
```

Reply:
I don't understand why in this case you would like to optimize the code. There are several parts in the code where we could do the same thing, but we concluded that given the small number of constructors, it was unnecessary. Besides, we're losing here in concision. 
Moreover, I'm doing a conversion to sets here, but why? The function could accept lists, couldn't it? Is there another reason behind to use this function elsewhere?

Reply:
There are 2 optims here. We go through the list of constructors only once, and we only record the names starting with 'x' or 'p'. There are perhaps other places we could optim but it doesn't cost much to do it here. Moreover, in some applications, there can be hundreds of constructors. So, thanks to change the code.

Reply:
OK done